source/common/network/io_uring_socket_handle_impl.cc

```diff
@@ -0,0 +1,471 @@
+  is_read_added_ = false;
+
+  uint64_t len = bytes_to_read_;
+  bytes_to_read_ = 0;
+  return Api::IoCallUint64Result(len, Api::IoErrorPtr(nullptr, IoSocketError::deleteIoError));
+}
+
+Api::IoCallUint64Result IoUringSocketHandleImpl::writev(const Buffer::RawSlice* /*slices */,
+                                                        uint64_t /*num_slice*/) {
+  PANIC("not implemented");
+}
+
+Api::IoCallUint64Result IoUringSocketHandleImpl::write(Buffer::Instance& buffer) {
+  auto length = buffer.length();
+  ASSERT(length > 0);
+
+  std::list<Buffer::SliceDataPtr> slices;
+  while (buffer.length() > 0) {
+    // The buffer must not own the data after it has been extracted and put into
+    // the `io-uring` submission queue to avoid freeing it before the writev
+    // operation is completed.
+    Buffer::SliceDataPtr data = buffer.extractMutableFrontSlice();
+    slices.push_back(std::move(data));
+  }
+
+  uint32_t nr_vecs = slices.size();
+  struct iovec* iovecs = new struct iovec[slices.size()];
+  struct iovec* iov = iovecs;
+  for (auto& slice : slices) {
+    absl::Span<uint8_t> mdata = slice->getMutableData();
+    iov->iov_base = mdata.data();
+    iov->iov_len = mdata.size();
+    iov++;
+  }
+
+  auto req = new Request{*this, RequestType::Write, iovecs, std::move(slices)};
+  auto& uring = io_uring_factory_.getOrCreate();
+  auto res = uring.prepareWritev(fd_, iovecs, nr_vecs, 0, req);
+  if (res == Io::IoUringResult::Failed) {
+    // TODO(rojkov): handle `EBUSY` in case the completion queue is never reaped.
+    uring.submit();
+    res = uring.prepareWritev(fd_, iovecs, nr_vecs, 0, req);
+    RELEASE_ASSERT(res == Io::IoUringResult::Ok, "unable to prepare writev");
+  }
+  return Api::IoCallUint64Result(length, Api::IoErrorPtr(nullptr, IoSocketError::deleteIoError));
+}
+
+Api::IoCallUint64Result
+IoUringSocketHandleImpl::sendmsg(const Buffer::RawSlice* /*slices*/, uint64_t /*num_slice*/,
+                                 int /*flags*/, const Address::Ip* /*self_ip*/,
+                                 const Address::Instance& /*peer_address*/) {
+  PANIC("not implemented");
+}
+
+Api::IoCallUint64Result IoUringSocketHandleImpl::recvmsg(Buffer::RawSlice* /*slices*/,
+                                                         const uint64_t /*num_slice*/,
+                                                         uint32_t /*self_port*/,
+                                                         RecvMsgOutput& /*output*/) {
+  PANIC("not implemented");
+}
+
+Api::IoCallUint64Result IoUringSocketHandleImpl::recvmmsg(RawSliceArrays& /*slices*/,
+                                                          uint32_t /*self_port*/,
+                                                          RecvMsgOutput& /*output*/) {
+  PANIC("not implemented");
+}
+
+Api::IoCallUint64Result IoUringSocketHandleImpl::recv(void* /*buffer*/, size_t /*length*/,
+                                                      int /*flags*/) {
+  PANIC("not implemented");
```

Comment:
I do not see and equivalent operations of `flags` in io_uring like `MSG_PEEK`. We may implement on our own.